### ISSUE TITLE
clarify the intent of a complete table of contents

### ISSUE THREAD

**Issue by mattgarrish** [w3c/epub-specs#666]
The toc nav currently says this:

For usability and accessibility reasons, Authors should provide a comprehensive table of contents: the toc nav should not exclude references based solely on their nesting depth within the document hierarchy, as is often the case in print works (particularly in reduced tables of contents).

What "comprehensive" means here is not very clear. We should revisit and clarify. For example, it might help to not muddy the toc with structured sidebars and the like by changing to "a table of contents that includes references to all sections that contribute to the primary reading order of the EPUB Publication"

**Comment by mattgarrish** [comment]
Also problematic for validation is that the toc link can't always directly reference a heading element since there could be content before it (i.e., it's usually the pointer to the start of the section using the heading as the name).

But if it references a sectioning element, we don't require any explicit linking between the element and its heading (either via aria-label or aria-labelledby), so it requires an assumption that the first heading element found is the heading (role="heading" muddies what is a heading, too).

Also, it's not always the case that the toc name and the section heading exactly match, so without explicit identifiers it's hard to be sure that a heading matches the toc, or if there is a mismatch that needs reporting.

Long story short, there might be ways to address these issues through better spec prose, but it also might increase the production overhead.

**Comment by rdeltour** [comment]
> Also problematic for validation is that the toc link can't always directly reference a heading element since there could be content before it (i.e., it's usually the pointer to the start of the section using the heading as the name).
> 
> But if it references a sectioning element, we don't require any explicit linking between the element and its heading (either via aria-label or aria-labelledby), so it requires an assumption that the first heading element found is the heading (role="heading" muddies what is a heading, too).

I'm not sure I'm understanding the point here. Are you suggesting that the `toc nav` constraints s/b strengthened by mandating the link to point to either a _sectioning content_ or a _heading_?

If not, the pain point of validation is more general that the points above: it's about relating any fragment to its containing section.

> Also, it's not always the case that the toc name and the section heading exactly match, so without explicit identifiers it's hard to be sure that a heading matches the toc, or if there is a mismatch that needs reporting.
> 
> Long story short, there might be ways to address these issues through better spec prose, but it also might increase the production overhead.

IMO the only viable approach to make the rule precise without muddying it too much is to clarify its relation with the HTML outline; notably how the outlines of spine docs are combined.  
If we don't want to follow that road, there's nothing else we can do, and the _SHOULD_ there will never be fully deterministically testable. 
I like the idea of mentioning the "primary reading order", but wouldn't it need to be properly defined then?

**Comment by mattgarrish** [comment]
> notably how the outlines of spine docs are combined

That's another flaw in the definition. What does "aggregation of outlines" mean when each document is a distinct entity and the outlines always are forced to begin from an h1?

Without trying to write a new model to concatenate -- and I'm not even sure that's realistic so long as we have fixed layout publications, even if we could exclude non-linear content -- pinning the toc entries more concretely to the content seems the only other way forward, but I think it's equally fatal for validation.

For example, is it enough just to point to an ID of something and assume that because that something is there it's still a valid reference? If we want more stringent checking, I think we're bound to fail.

EPUB content doesn't have to be structured like it does in epub, so I think we're at the same place as far as fragments. What I was trying to note is that is there a problem if content precedes a heading. Is it the heading for the containing sectioning element (the body) or does it represent a section of content missing a heading? If we don't always have explicit linking from sectioning element to heading, there will always be ambiguous cases.

I'm thinking that solving some of these issues might be more realistically done in an accessibility profile where we could impose sectioning requirements like in edupub, but even then I think we might always have to leave some ambiguity.

> but wouldn't it need to be properly defined then?

Yes, I think that's part of the needed guidance for clarifying "completeness", but I'm not sure how we write it. Is content in an article part of the primary reading order or not? In a magazine, yes, but generally its structure is independent of the reading order.

**Comment by rdeltour** [comment]
> What I was trying to note is that is there a problem if content precedes a heading. Is it the heading for the containing sectioning element (the body) or does it represent a section of content missing a heading? 

I don't think there is any ambiguity here, as the association from section to heading content is well defined in HTML, precisely in the `heading content` and `outline` concepts.

The case where a section has no heading content, is OK too, as long as:
- we don't mandate `toc` links content to be equal to the section's heading.
- we don't mandate `toc` links to point to a section's heading

> If we don't always have explicit linking from sectioning element to heading, there will always be ambiguous cases.

Why, can you elaborate?

**Comment by mattgarrish** [comment]
It's defined as theory, sure, but as practice? I don't know how realistic it is that validation could be pinned to it, if for no other reason than the lack of awareness or understanding. How many people can even read the outline algorithm and correctly predict the edge cases? Given the way content gets authored, it will lead to frustrating to solve incomplete table of contents reports.

I'm also not seeing how you're going to merge the result of independent document algorithms into a meaningful publication outline without more information. The spine is a flat list and it is not required that each chunk start with a top-level heading, so what is the authoritative source for how you combine them? How do you know which fxl pages aren't expected to have headings in them?

That's where I was going with explicit linking. If we are only out to validate the table of contents entries, and not go near the outline algorithm, explicit linking would always take you from the reference in the toc to the heading in the content without any need to build an outline. Any heading that can't be backtracked could be flagged; similarly any heading level that doesn't match the nesting level in the toc. You might get to the same place with some heuristics, but it takes the guessing out.

I was also looking at what's needed for WCAG compliance. Completeness is not a requirement, but you are expected to:

> Check that the values and order of the entries in the table of contents correspond to the names and order of the sections of the document.
> https://www.w3.org/TR/WCAG20-TECHS/G64.html

You can verify ordering is sequential easily enough, but if names are not the same are we risk moving into into false-positive territory. Differences between toc and content aren't uncommon (truncation for example), so how do you tell a minor wording variation from a toc entry that simply doesn't match the content anymore?

**Comment by rdeltour** [comment]
> It's defined as theory, sure, but as practice? I don't know how realistic it is that validation could be pinned to it, if for no other reason than the lack of awareness or understanding. How many people can even read the outline algorithm and correctly predict the edge cases? Given the way content gets authored, it will lead to frustrating to solve incomplete table of contents reports.

Oh I completely agree.  
What I was trying to say is that if we ever want to actually check the toc exhaustiveness we have to either:
- use the semantics defined in the outline algorithm
- OR, somehow constrain to the Content Doc structure to make the outline more explicit e.g. "sections must be explicitly defined with a `section` element".

For the reasons you describe (basically that _outline_ is not implemented by browsers or ATs and thus poorly understood by the common web developer), I believe my first bullet above is unrealistic. The second bullet essentially boils down to doing what EDUPUB does, which is out of scope for the generic EPUB profile.

> I'm also not seeing how you're going to merge the result of independent document algorithms into a meaningful publication outline without more information. The spine is a flat list and it is not required that each chunk start with a top-level heading, so what is the authoritative source for how you combine them? How do you know which fxl pages aren't expected to have headings in them?

Yes, combining outlines would definitely require more information, like a `depth` attribute on the spine items.

> That's where I was going with explicit linking. If we are only out to validate the table of contents entries, and not go near the outline algorithm, explicit linking would always take you from the reference in the toc to the heading in the content without any need to build an outline.

Do you agree that would require changes to the `toc` links definition? Unless we say `toc` links **must** point to heading content or say heading-less sectioning content, I don't see what you mean by "explicit linking". (see also my first question in [my first comment above](#issuecomment-180101044)).

> Any heading that can't be backtracked could be flagged; similarly any heading level that doesn't match the nesting level in the toc. You might get to the same place with some heuristics, but it takes the guessing out.

That could work **only if** we add further requirements on `toc` links (e..g they must point to headings). With the spec as it is currently written, we cannot backtrack a heading without relying on sectioning semantics, i.e. the outline algorithm.

See for instance:

``` xml
<section>
     <span id="sect-42"/>
     <div class="beautiful-heading">
       <h3>...</h3>
     </div>
```

How to backtrack the `h3` to a `toc link` that would point to `#sect-42` without relying on the sectioning semantics?

> I was also looking at what's needed for WCAG compliance. Completeness is not a requirement, but you are expected to:
> 
> Check that the values and order of the entries in the table of contents correspond to the names and order of the sections of the document.
> https://www.w3.org/TR/WCAG20-TECHS/G64.html
> You can verify ordering is sequential easily enough, but if names are not the same are we risk moving into into false-positive territory. Differences between toc and content aren't uncommon (truncation for example), so how do you tell a minor wording variation from a toc entry that simply doesn't match the content anymore?

Yes, I think that's one area where automated conformance checking is not enough.

**Comment by mattgarrish** [comment]
By explicit I mean we would require either pointing directly to a heading or, if you point to a sectioning element that that element either has an aria-label or aria-labelledby.

So your example would have to be:

``` html
<section aria-labelledby="sec-hd">
     <span id="sect-42"/>
     <div class="beautiful-heading">
       <h3 id="sec-hd">...</h3>
     </div>
```

Matching toc entry to heading becomes unambiguous regardless of whether you're crawling the toc or the content. It doesn't solve the problem of completeness, but I don't think we can easily solve that. We would need stronger sectioning rules like in edupub first so that each heading can only be tied to one section.

It's more authoring work to do this, too, but it's technically the way you're supposed to link a heading to its section if you want to have an aria landmark for the section (unlabelled sections without a role not appearing in the landmark list by default).

**Comment by mattgarrish** [comment]
I'm thinking we should strip the second and third paragraph of the toc nav definition and just leave this prose:

> The toc nav element defines the primary navigational hierarchy of the given Rendition. It conceptually corresponds to a table of contents in a printed work (i.e., it provides navigation to the major structural sections of the publication).
> The order of li elements contained within the toc nav element must match the order of the targeted elements within each targeted EPUB Content Document, and must also follow the order of Content Documents in the Rendition's spine.
> The toc nav element must occur exactly once in an EPUB Navigation Document.

We can continue to discuss whether there is a meaningful definition of what has to be in a table of contents for accessibility, and add to the accessibility techniques, if so.

**Comment by laudrain** [comment]
I agree.

**Comment by TzviyaSiegman** [comment]
+1

**Comment by rdeltour** [comment]
> I'm thinking we should strip the second and third paragraph of the toc nav definition and just leave this prose

LGTM.

**Comment by mattgarrish** [comment]
Proposed solution:

Remove the second and third paragraphs, leaving the description as:

The toc nav element defines the primary navigational hierarchy of the given Rendition. It conceptually corresponds to a table of contents in a printed work (i.e., it provides navigation to the major structural sections of the publication).

The order of li elements contained within the toc nav element must match the order of the targeted elements within each targeted EPUB Content Document, and must also follow the order of Content Documents in the Rendition's spine.

The toc nav element must occur exactly once in an EPUB Navigation Document.

**Comment by mattgarrish** [comment]
The two paragraphs have been removed from the definition:
https://rawgit.com/IDPF/epub-revision/master/build/31/spec/epub-packages.html#sec-nav-toc

Please close this issue if no other changes are required (with the understanding that this change could be revisited later by the a11y subgroup, but would be in a new issue).
